File: 序号/2251. 花期内花的数目.kt
class Solution {
  
  fun fullBloomFlowers(flowers: Array<IntArray>, persons: IntArray): IntArray {
    // 先用差分记录花开花落，再用前缀和记录几天前开了多少花，最后用二分求特定时间点花开数量
    val map = TreeMap<Int,Int>() // key 表示时间，value 表示开花或落花的数目
    for((start,end) in flowers){
      map[start] =  map.getOrDefault(start,0)+1
      map[end+1] = map.getOrDefault(end+1,0) -1 // end那一天花还是开着的，所以要+1
    }
    val saki = TreeMap<Int,Int>() // key 表示时间，value 表示开花的数量
    var cnt = 0
    for((time,n) in map){
      cnt+=n
      saki[time]=cnt
    }
    val n = persons.size
    val ans = IntArray(n)
    for(i in persons.indices){
      var entry = saki.floorEntry(persons[i])
      if(entry!=null && entry.key==persons[i]){
        ans[i]=entry.value
        continue
      }
      entry = saki.lowerEntry(persons[i])
      if(entry!=null){
        ans[i]=entry.value
      }
    }
    return ans
  }
}

File: 序号/2251. 花期内花的数目.py
class Solution:
    def fullBloomFlowers(self, flowers: List[List[int]], people: List[int]) -> List[int]:
        # flowers[i] 每朵花开放时间  people 人来的时间  返回 人到的时候花的开放数目
        # 离散 差分 前缀和
        times={50001}
        for l,r in flowers:
            times.add(l)
            times.add(r+1)
        sortedTime = sorted(times)
        timeDict={}
        for i,time in enumerate(sortedTime):
            timeDict[time]=i
        diffList=[0]*len(sortedTime)
        for l,r in flowers:
            diffList[timeDict[l]]+=1
            diffList[timeDict[r+1]]-=1
        for i in range(1,len(diffList)):
            diffList[i]+=diffList[i-1]
        ans=[0]*len(people)
        for i,p in enumerate(people):
            l,r=-1,len(diffList)
            while l+1<r:
                mid=(l+r)//2
                acturlTime=sortedTime[mid]
                if acturlTime<=p:
                    l=mid
                elif acturlTime>p:
                    r=mid
            ans[i]=diffList[l]
        return ans